format/puffin-spec.md

```diff
@@ -0,0 +1,144 @@
+- `Footer` is defined below.
+
+### Footer structure
+
+Footer has the following structure
+
+```
+Magic FooterPayload FooterPayloadSize Flags Magic
+```
+
+where
+
+- `Magic`: four bytes, same as at the beginning of the file
+- `FooterPayload`: optionally compressed, UTF-8 encoded JSON payload describing the
+  blobs in the file, with the structure described below
+- `FooterPayloadSize`: a length in bytes of the `FooterPayload` (after compression,
+  if compressed), stored as 4 byte integer
+- `Flags`: 4 bytes for boolean flags
+  - byte 0 (first)
+    - bit 0 (lowest bit): whether `FooterPayload` is compressed
+    - all other bits are reserved for future use and should be set to 0 on write
+  - all other bytes are reserved for future use and should be set to 0 on write
+
+A 4 byte integer is always signed, in a two's complement representation, stored
+little-endian.
+
+### Footer Payload
+
+Footer payload bytes is either uncompressed or LZ4-compressed (as a single
+[LZ4 compression frame](https://github.com/lz4/lz4/blob/77d1b93f72628af7bbde0243b4bba9205c3138d9/doc/lz4_Frame_format.md)
+with content size present), UTF-8 encoded JSON payload representing a single
+`FileMetadata` object.
+
+#### FileMetadata
+
+`FileMetadata` has the following fields
+
+
+| Field Name | Field Type                              | Required | Description |
+| ---------- | --------------------------------------- | -------- | ----------- |
+| blobs      | list of BlobMetadata objects            | yes      |
+| properties | JSON object with string property values | no       | storage for arbitrary meta-information, like writer identification/version. See [Common properties](#common-properties) for properties that are recommended to be set by a writer.
+
+#### BlobMetadata
+
+`BlobMetadata` has the following fields
+
+| Field Name        | Field Type                              | Required | Description |
+|-------------------|-----------------------------------------|----------| ----------- |
+| type              | JSON string                             | yes      | See [Blob types](#blob-types)
+| fields            | list of JSON long                       | yes      | List of field IDs the blob was computed for; the order of items is used to compute sketches stored in the blob.
+| offset            | JSON long                               | yes      | The offset in the file where the blob contents start
+| length            | JSON long                               | yes      | The length of the blob stored in the file (after compression, if compressed)
+| compression-codec | JSON string                             | no       | See [Compression codecs](#compression-codecs). If omitted, the data is assumed to be uncompressed.
+| properties        | JSON object with string property values | no       | storage for arbitrary meta-information about the blob
+
+### Blob types
+
+The blobs can be of a type listed below
+
+#### `apache-datasketches-theta-v1` blob type
```

Comment:
So to clarify, alpha-sketch is just one example, and the idea here is it support various statistics/secondary index that is to be defined, right?  Example, other sketches, or even bloom filter?